### ISSUE TITLE
Petals doesn't deal with server failure properly

### ISSUE THREAD

**Issue by oldcpple** [bigscience-workshop/petals#587]
Hi there, we'd like to report our findings on testing Petals' availability of fault tolerance.

We note that the current implementation of the method _step_ in the class __ServerInferenceSession_ from _inference_session.py_ contains the following content:
`if self.history is None:
    self.history = inputs
elif self.history.shape[1] == self._position:
    self.history = torch.cat([self.history, inputs[:, -n_input_tokens:]], dim=1)

assert self.history.shape[1] == self._position + n_input_tokens, (
    f"Broken input cache: span={self.span} shape={self.history.shape} "
    f"position={self._position} n_input_tokens={n_input_tokens}"
)
`
where the attributes self.history and self.position are initialized as None and 0 respectively when an object of the class __ServerInferenceSession_ is created. The problem is, when a server fails, Petals replaces it with another server that serves the same blocks. However, the new server session is just initialized when joining the inference session, and its attribute _position_ is 0.

In the method __update_sequence_ of the class _InferenceSession_, the new server session's _history_ will be assigned the history of the failed server session:
`updated_sessions[0].history = self._server_sessions[server_idx].history
`
And during the inference, _n_input_tokens_ will always be 1. Thus, the assertion:
`assert self.history.shape[1] == self._position + n_input_tokens
`
is always likely to throw exceptions of "Broken input cache".

One possible solution is to update the attribute _position_ as well. For example, change in the method __update_sequence_:
`if server_idx < n_prev_spans:
    updated_sessions[0].history = self._server_sessions[server_idx].history
    updated_sessions[0]._position = self._server_sessions[server_idx]._position
`

It does work in our tests. When a server fails and Petals selects another, the inference procedure continues without the aforementioned exception. However, the results could be incorrect. For instance, for the input prompt "How do you build a strong network?", without server failure the model would generate the output "A: Building a strong network is essential for success in any field. Here are some tips to help you build a strong network...".

With the current Petals implementation, the recovery process will encounter exceptions as mentioned above. With our change, Petals recovers the inference session, but the output would be: "A: Building a strong network is essential, but it's not always easy. Here are some tips to help you build strong relationships with your colleagues...".

Please check if there are such problems. Many thanks.

**Comment by justheuristic** [comment]
@oldcpple Thank you! (and sorry for the delayed response)

Your reasoning looks sound. If you still have bandwidth, would you kindly provide a minimal example where this assert fails? (so we can reproduce it for CI tests) That would ensure that, once fixed, this error does not resurface.
I tried to check this via debugprints using our mini-swarm from CI, by shutting down servers during inference, but assert would evaluate true all of 10-ish times it tested it.

If you don't, please reply as such, and I will continue trying.

**Comment by justheuristic** [comment]
Discussed this with @borzunov (core author) : he also believes that you are correct and hopes to make a pass this weekend (please note that this is not a binding promise, we are unfortunately still overwhelmed with research routine).

**Comment by oldcpple** [comment]
Thanks for your reply! 
We were running a StableBeluga2 model on a 10×2080Ti private swarm, and made sure that each block of the model was held by at least 2 servers. We then disconneted one of a servers on the server chain selected by the client,  it would take a period of time to find another server to replace the disconneted one, and that was when the assert failed.
We tried to attribute the reason and simply fixed it with a few lines of code, as described above. Now the system can successfully recover the interrupted inference can the final result is the same as the one where no server disconnected. But we haven't varified whether the change we made would affect other parts.

As you mentioned that your tests ended with correct results, we guess it might be a model-specified problem (but we haven't varified this). Note that we met this problem with StableBeluga2, we further found another problem with this model:  the batch size seems to be fixed. For example, we first tried to send a request with 2 prompts(batch size=2), the procedure of inference was fine. Then for the following requests, we changed batch size to 10 or any other except 2, similar "Broken input cache" exceptions would occured. We tested for a few times, and was sure that the batch size the model can process must be the same as the batch size of the first inference request. But we believe this as a problem of StableBeluga2, since everything was fine in our test on another model, bloom-7b.

**Comment by justheuristic** [comment]
For the record: the issue certainly exists and I can confirm that it breaks fault tolerance on my side. We have not yet fixed it as the issue is bouncing between myself and @borzunov promising to "certainly fix this next weekend night". We hope to fix it and post an update, but everything takes unreasonably long at the moment. We are sorry.
